voxel_grid/prototype/sfinae_check.hpp

```diff
@@ -0,0 +1,136 @@
+// #include <iostream>
+#include <boost/optional.hpp>
+#include <type_traits>
+#include <vector>
+
+template <typename, typename>
+class TransformFilter;
+
+struct PointXYZ {};
+
+namespace detail {
+using std::begin;
+
+template <typename C, typename PointT>
+struct IsGridStruct {
+ private:
+  template <typename T>
+  static constexpr auto hasSetUp(T*) ->
+      typename std::is_same<decltype(std::declval<T>().setUp(
+                                std::declval<TransformFilter<T, PointT>*>())),
+                            bool>::type;
+  template <typename>
+  static constexpr std::false_type hasSetUp(...);
+
+  template <typename T>
+  static constexpr auto hasAddPoint(T*) -> typename std::is_same<
+      decltype(std::declval<T>().addPointToGrid(std::declval<PointT&>())),
+      void>::type;
+  template <typename>
+  static constexpr std::false_type hasAddPoint(...);
+
+  template <typename T>
+  static constexpr auto hasFilterBegin(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(begin(std::declval<T>()))),
+      boost::optional<PointT>>::type;
+  template <typename T>
+  static constexpr auto hasFilterBegin(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(std::declval<T>().begin())),
+      boost::optional<PointT>>::type;
+  template <typename>
+  static constexpr std::false_type hasFilterBegin(...);
+
+  template <typename T>
+  static constexpr auto hasFilterEnd(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(end(std::declval<T>()))),
+      boost::optional<PointT>>::type;
+  template <typename T>
+  static constexpr auto hasFilterEnd(T*) -> typename std::is_same<
+      decltype(std::declval<T>().filterGrid(std::declval<T>().end())),
+      boost::optional<PointT>>::type;
+  template <typename>
+  static constexpr std::false_type hasFilterEnd(...);
+
+  template <typename T>
+  static constexpr auto hasSize(T*) ->
+      typename std::is_same<decltype(std::declval<T>().size()),
+                            std::size_t>::type;
+  template <typename>
+  static constexpr std::false_type hasSize(...);
+
+  using has_set_up = decltype(hasSetUp<C>(0));
+  using has_add_point = decltype(hasAddPoint<C>(0));
+  using has_filter_begin = decltype(hasFilterBegin<C>(0));
+  using has_filter_end = decltype(hasFilterEnd<C>(0));
+  using has_size = decltype(hasSize<C>(0));
+
+ public:
+  static constexpr bool is_set_up_valid = has_set_up::value;
+  static constexpr bool is_add_point_valid = has_add_point::value;
+  static constexpr bool is_filter_valid =
+      has_filter_begin::value && has_filter_end::value;
+  static constexpr bool is_size_valid = has_size::value;
+
+  static constexpr bool is_valid =
+      is_set_up_valid && is_add_point_valid && is_filter_valid && is_size_valid;
+};
+}  // namespace detail
+
+template <typename GridStruct>
```

Comment:
Can you also add an untemplated function just for satisfaction

Reply:
Like this?

https://github.com/tin1254/PCL_GSoC_2021/blob/013b37617e57a302e65d27a62bdd1e82fc6a9a25/voxel_grid/prototype/sfinae_check.hpp#L87-L91